lib/models/per-bundle-addon-cache/index.js

```diff
@@ -0,0 +1,276 @@
+    this.projectAddonCache = new Map();
+
+    // cache of the addon packageInfos for addons that are descendants of the project
+    // (excluding going into lazy engines). Lazy engines that refer to any of these
+    // will defer to the project-level one.
+    this.bundledProjectAddonPkgInfos = this.setupBundledProjectAddonPackageInfos();
+
+    // keyed by engine name, the name, engine instance and addon instance cache 
+    // (itself keyed by addon realPath) for each engine.
+    this.lazyEngineCache = this.setupLazyEngineCache();
+  }
+
+  /**
+     * Get the list of addon PackageInfo objects that are a dependency of this project,
+     * directly or transitively. Do not check any dependency that is itself a bundle root.
+     * This is used as part of the per-bundle addon caching optimization.
+     *
+     * @method getBundledAddonPackageInfos
+     * @return {Set} a Set of {PackageInfo} objects.
+     */
+  setupBundledProjectAddonPackageInfos() {
+    debugger;
+    let pkgInfos = new Set();
+
+    let findAddons = (pkgInfo) => {
+      if (!pkgInfo.valid || !pkgInfo.addonMainPath) {
+        return;
+      }
+
+      if (pkgInfos.has(pkgInfo)) {
+        return;
+      }
+
+      if (pkgInfo.isLazyEngine()) {
+        return;
+      }
+
+      pkgInfos.add(pkgInfo);
+
+      let addonPackageList = pkgInfo.discoverAddonAddons();
+      addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+    }
+
+    let addonPackageList = this.project._packageInfo.discoverProjectAddons();
+    addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+
+    return pkgInfos;
+  }
+  
+  /**
+  * Find all lazy engine packageInfos in the project. We cache only a single copy
+  * of each, by NAME, not by PATH.
+  * 
+  * @public
+  * @method setupBundleRoots
+  * @param {Project} project the project that we use to get the PackageInfoCache
+  */
+  setupLazyEngineCache() {
+    debugger;
+    let lazyEnginePkgInfos = new Map();
+
+    let findAddons = (pkgInfo) => {
+      if (!pkgInfo.valid || !pkgInfo.addonMainPath) {
+        return;
+      }
+
+      if (pkgInfo.isLazyEngine()) {
+        if (lazyEnginePkgInfos.has(pkgInfo.name)) {
+          return;
+        }
+
+        lazyEnginePkgInfos.set(pkgInfo.name, pkgInfo);
+      }
+
+      let addonPackageList = pkgInfo.discoverAddonAddons();
+      addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+    }
+
+    let addonPackageList = this.project._packageInfo.discoverProjectAddons();
+    addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+
+    //   engineName: xxx
+    //   engineInstance: yyy (initially null, filled in as we instantiate an engine
+    //   addonInstanceCache: zzz (the cache of addon instances and proxies bundled with this lazy engine))
+    let cache = new Map();
+    lazyEnginePkgInfos.forEach((pkgInfo) => {
+      cache.set(pkgInfo.name, new Object({engineName: pkgInfo.name, engineInstance: null, addonInstanceCache: new Map()}));
```

Comment:
```suggestion
      cache.set(pkgInfo.name, {engineName: pkgInfo.name, engineInstance: null, addonInstanceCache: new Map()});
```

Reply:
Done. As I'm changing my own code directly, won't commit the suggestion itself.